bindgen/bindgen.bzl

```diff
@@ -0,0 +1,167 @@
+
+    rust_bindgen(
+        name = name + "__bindgen",
+        header = header,
+        cc_lib = cc_lib,
+        **kwargs
+    )
+    rust_library(
+        name = name,
+        srcs = [name + "__bindgen.rs"],
+        deps = [cc_lib],
+    )
+
+def _rust_bindgen_impl(ctx):
+    # nb. We can't grab the cc_library`s direct headers, so a header must be provided.
+    cc_lib = ctx.attr.cc_lib
+    header = ctx.file.header
+    if header not in cc_lib.cc.transitive_headers:
+        fail("Header {} is not in {}'s transitive headers.".format(ctx.attr.header, cc_lib), "header")
+
+    toolchain = ctx.toolchains["@io_bazel_rules_rust//bindgen:bindgen_toolchain"]
+    bindgen_bin = toolchain.bindgen
+    rustfmt_bin = toolchain.rustfmt
+    clang_bin = toolchain.clang
+    libclang = toolchain.libclang
+
+    # TODO: This should not need to be explicitly provided, see below TODO.
+    libstdcxx = toolchain.libstdcxx
+
+    # rustfmt is not where bindgen expects to find it, so we format manually
+    bindgen_args = ["--no-rustfmt-bindings"] + ctx.attr.bindgen_flags
+    clang_args = ctx.attr.clang_flags
+
+    output = ctx.outputs.out
+
+    # libclang should only have 1 output file
+    libclang_dir = libclang.cc.libs.to_list()[0].dirname
+    include_directories = depset(
+        [f.dirname for f in cc_lib.cc.transitive_headers],
+    )
+
+    if rustfmt_bin:
+        unformatted_output = ctx.actions.declare_file(output.basename + ".unformatted")
+    else:
+        unformatted_output = output
+
+    args = ctx.actions.args()
+    args.add_all(bindgen_args)
+    args.add(header.path)
+    args.add("--output", unformatted_output.path)
+    args.add("--")
+    args.add_all(include_directories, before_each = "-I")
+    args.add_all(clang_args)
+    ctx.actions.run(
+        executable = bindgen_bin,
+        inputs = depset(
+            [header],
+            transitive = [cc_lib.cc.transitive_headers, libclang.cc.libs, libstdcxx.cc.libs],
+        ),
+        outputs = [unformatted_output],
+        mnemonic = "RustBindgen",
+        progress_message = "Generating bindings for {}..".format(header.path),
+        env = {
+            "RUST_BACKTRACE": "1",
+            "CLANG_PATH": clang_bin.path,
+            # Bindgen loads libclang at runtime, which also needs libstdc++, so we setup LD_LIBRARY_PATH
+            "LIBCLANG_PATH": libclang_dir,
+            # TODO: If libclang were built by bazel from source w/ properly specified dependencies, it
+            #       would have the correct rpaths and not require this nor would this rule
+            #       have a direct dependency on libstdc++.so
+            #       This is unnecessary if the system libstdc++ suffices, which may not always be the case.
+            "LD_LIBRARY_PATH": ":".join([f.dirname for f in libstdcxx.cc.libs]),
+        },
+        arguments = [args],
+        tools = [clang_bin],
+    )
+
+    if rustfmt_bin:
```

Comment:
Is running rustfmt needed on generated sources? I meant, they are not made for being read anyway.

Reply:
I've personally found it useful (in my own hacky version of this) when inspecting generated files. In my experience, bindgen can be a bit finnicky and require special params to configure generation.

Reply:
Beyond being useful, this is what bindgen does by default. I think it is best to keep parity with other tools whenever possible.

Reply:
Sounds good, can you just add a comment about it?